File: packages/cpp/examples/provider_stream_no_key.cpp
#include "axllm/axllm.hpp"
#include <iostream>
#include <string>

struct LegacyTransport : axllm::Transport {
  axllm::Value call(axllm::Value) override {
    return axllm::object({
      {"status", 200},
      {"body",
       "data: {\"id\":\"chatcmpl_stream\",\"model\":\"gpt-5.4-mini\",\"choices\":[{\"index\":0,\"delta\":{\"content\":\"hel\"}}]}\n\n"
       "data: {\"id\":\"chatcmpl_stream\",\"model\":\"gpt-5.4-mini\",\"choices\":[{\"index\":0,\"delta\":{\"content\":\"lo\"},\"finish_reason\":\"stop\"}]}\n\n"
       "data: {\"id\":\"chatcmpl_stream\",\"model\":\"gpt-5.4-mini\",\"choices\":[],\"usage\":{\"prompt_tokens\":1,\"completion_tokens\":2,\"total_tokens\":3}}\n\n"
       "data: [DONE]\n\n"}
    });
  }

};

struct IncrementalTransport : axllm::Transport {
  bool cancelled = false;
  axllm::Value call(axllm::Value) override { return axllm::Value::object(); }
  void stream(axllm::Value, axllm::AxTransportStreamHandler handler) override {
    const std::string body =
      "data: {\"id\":\"chatcmpl_cancel\",\"model\":\"gpt-5.4-mini\",\"choices\":[{\"index\":0,\"delta\":{\"content\":\"hel 🌍\"}}]}\r\n\r\n"
      "data: {\"id\":\"chatcmpl_cancel\",\"model\":\"gpt-5.4-mini\",\"choices\":[{\"index\":0,\"delta\":{\"content\":\"lo\"}}]}\r\n\r\n";
    for (char byte : body) {
      if (!handler(std::string(1, byte))) {
        cancelled = true;
        return;
      }
    }
  }
};

struct FailingTransport : axllm::Transport {
  int attempts = 0;
  axllm::Value call(axllm::Value) override { return axllm::Value::object(); }
  void stream(axllm::Value, axllm::AxTransportStreamHandler handler) override {
    ++attempts;
    handler("data: {\"id\":\"chatcmpl_failure\",\"model\":\"gpt-5.4-mini\",\"choices\":[{\"index\":0,\"delta\":{\"content\":\"delivered\"}}]}\n\n");
    throw axllm::AxError("network", "upstream closed", "AxAIServiceNetworkError", 0, "", true);
  }
};

int main() {
  LegacyTransport legacy;
  axllm::OpenAICompatibleClient client(axllm::object({{"api_key", "test-key"}, {"model", "gpt-5.4-mini"}}), &legacy);
  std::string text;
  std::vector<axllm::AxUsageEvent> usage_events;
  axllm::set_usage_observer([&](axllm::AxUsageEvent event) { usage_events.push_back(event); });
  for (const auto& event : client.stream(axllm::object({
         {"chat_prompt", axllm::array({axllm::object({{"role", "user"}, {"content", "stream"}})})}
       }))) {
    text += axllm::display(axllm::Core::get(axllm::Core::get(axllm::Core::get(event, "results"), 0), "content", ""));
  }
  axllm::set_usage_observer({});
  if (text != "hello") return 1;
  if (usage_events.size() != 1) return 4;
  IncrementalTransport incremental;
  axllm::OpenAICompatibleClient cancel_client(axllm::object({{"api_key", "test-key"}, {"model", "gpt-5.4-mini"}}), &incremental);
  cancel_client.stream_each(axllm::object({
      {"chat_prompt", axllm::array({axllm::object({{"role", "user"}, {"content", "cancel"}})})}
    }), [](const axllm::Value&) { return false; });
  if (!incremental.cancelled) return 2;
  FailingTransport failing;
  axllm::OpenAICompatibleClient failure_client(axllm::object({{"api_key", "test-key"}, {"model", "gpt-5.4-mini"}}), &failing);
  bool delivered = false;
  bool failed = false;
  try {
    failure_client.stream_each(axllm::object({
        {"chat_prompt", axllm::array({axllm::object({{"role", "user"}, {"content", "fail"}})})}
      }), [&](const axllm::Value&) { delivered = true; return true; });
  } catch (const axllm::AxError&) {
    failed = true;
  }
  if (!delivered || !failed || failing.attempts != 1) return 3;
  std::cout << "cpp-provider-stream-no-key " << text << "\n";
}

File: packages/python/examples/provider_stream_no_key.py
from axllm import AxAIServiceNetworkError, OpenAICompatibleClient, set_usage_observer


def scripted_transport(request):
    return {
        "status": 200,
        "body": (
            'data: {"id":"chatcmpl_stream","model":"gpt-5.4-mini","choices":[{"index":0,"delta":{"content":"hel"}}]}' + "\n\n"
            'data: {"id":"chatcmpl_stream","model":"gpt-5.4-mini","choices":[{"index":0,"delta":{"content":"lo"},"finish_reason":"stop"}]}' + "\n\n"
            'data: {"id":"chatcmpl_stream","model":"gpt-5.4-mini","choices":[],"usage":{"prompt_tokens":1,"completion_tokens":2,"total_tokens":3}}' + "\n\n"
            "data: [DONE]\n\n"
        ),
    }


client = OpenAICompatibleClient(api_key="test-key", model="gpt-5.4-mini", transport=scripted_transport)
usage_events = []
set_usage_observer(usage_events.append)
events = list(client.stream({"chat_prompt": [{"role": "user", "content": "stream"}]}))
set_usage_observer(None)
text = "".join((result.get("content") or "") for event in events for result in event["results"][:1])
assert text == "hello", events
assert len(usage_events) == 1, usage_events

closed = False
def incremental_transport(request):
    body = (
        'data: {"id":"chatcmpl_cancel","model":"gpt-5.4-mini","choices":[{"index":0,"delta":{"content":"first 🌍"}}]}\r\n\r\n'
        'data: {"id":"chatcmpl_cancel","model":"gpt-5.4-mini","choices":[{"index":0,"delta":{"content":"second"}}]}\r\n\r\n'
    ).encode()
    def chunks():
        global closed
        try:
            for byte in body:
                yield bytes([byte])
        finally:
            closed = True
    return chunks()

cancel_client = OpenAICompatibleClient(api_key="test-key", model="gpt-5.4-mini", transport=incremental_transport)
cancel_stream = cancel_client.stream({"chat_prompt": [{"role": "user", "content": "cancel"}]})
assert next(cancel_stream)["results"][0]["content"] == "first 🌍"
cancel_stream.close()
assert closed, "consumer cancellation did not close the upstream stream"

attempts = 0
def failing_transport(request):
    global attempts
    attempts += 1
    def chunks():
        yield b'data: {"id":"chatcmpl_failure","model":"gpt-5.4-mini","choices":[{"index":0,"delta":{"content":"delivered"}}]}\n\n'
        raise AxAIServiceNetworkError("upstream closed", retryable=True)
    return chunks()

failure_client = OpenAICompatibleClient(api_key="test-key", model="gpt-5.4-mini", transport=failing_transport)
failure_stream = failure_client.stream({"chat_prompt": [{"role": "user", "content": "fail"}]})
assert next(failure_stream)["results"][0]["content"] == "delivered"
try:
    next(failure_stream)
    raise AssertionError("mid-stream failure was not surfaced")
except AxAIServiceNetworkError:
    pass
assert attempts == 1, "mid-stream failure replayed the request"
print("python-provider-stream-no-key", text)
